x-pack/plugins/rule_registry/server/rule_data_client/index.ts

```diff
@@ -111,24 +113,86 @@ export class RuleDataClient implements IRuleDataClient {
     };
   }
 
+  createNamespacedIndexTemplate({
+    primaryNamespacedAlias,
+    secondaryNamespacedAlias,
+    namespace,
+  }: {
+    primaryNamespacedAlias: string;
+    secondaryNamespacedAlias?: string;
+    namespace?: string;
+  }): IndicesPutIndexTemplateRequest {
+    return {
+      name: primaryNamespacedAlias,
+      body: {
+        index_patterns: [`${primaryNamespacedAlias}-*`],
+        composed_of: [...this.options.componentTemplateNames],
+        template: {
+          aliases:
+            secondaryNamespacedAlias != null
+              ? {
+                  [secondaryNamespacedAlias]: {
+                    is_write_index: false,
+                  },
+                }
+              : undefined,
+          settings: {
+            'index.lifecycle': {
+              name: DEFAULT_ILM_POLICY_ID,
+              // TODO: fix the types in the ES package, they don't include rollover_alias???
+              // @ts-expect-error
+              rollover_alias: primaryNamespacedAlias,
+            },
+          },
+        },
+        _meta: {
+          namespace,
+        },
+        // By setting the priority to namespace.length, we ensure that if one namespace is a prefix of another namespace
+        // then newly created indices will use the matching template with the *longest* namespace
+        priority: namespace?.length,
```

Comment:
Not sure I got this part. Why do we need to set the priority if `namespace` cannot contain dashes?

I mean, let's say we have indices like

`.alerts-security.alerts-foo-000001`
`.alerts-security.alerts-foobar-000001`

Index template of the first one will have `index_patterns: ['.alerts-security.alerts-foo-*']` which will exclude the second index.

Sorry, maybe I'm just completely not seeing the idea.

Reply:
The namespace can contain dashes. Even if we prevent new namespaces from having dashes, we plan to use existing space IDs as namespaces for backwards compatibility, and space IDs can have dashes in them already.

Reply:
Oh... Holy moly, yes you're right!

Reply:
Here's an issue I wrote up on the problem with a little more detail: https://github.com/elastic/kibana/issues/107704

Proposal 4 on there is what I implemented here and I think it's definitely preferable to the other 3, but I had that idea after I already wrote the first 3 ideas so I just left them there